model/project.go

```diff
@@ -1019,62 +1002,25 @@ func (p *Project) FindDistroNameForTask(t *task.Task) (string, error) {
-	if projectRef == nil {
-		return nil, errors.New("projectRef given is nil")
+func FindLastKnownGoodProject(identifier string) (*Project, error) {
+	if identifier == "" {
+		return nil, errors.WithStack(errors.New("cannot pass empty identifier to FindLastKnownGoodProject"))
 	}
-	if projectRef.Identifier == "" {
-		return nil, errors.New("Invalid project with blank identifier")
+	project := &Project{
+		Identifier: identifier,
 	}
 
-	project := &Project{}
-	project.Identifier = projectRef.Identifier
-	// when the revision is empty we find the last known good configuration from the versions
-	// If the last known good configuration does not exist,
-	// load the configuration from the local config in the project ref.
-	if revision == "" {
-		lastGoodVersion, err := VersionFindOne(VersionByLastKnownGoodConfig(projectRef.Identifier))
-		if err != nil {
-			return nil, errors.Wrapf(err, "Error finding recent valid version for '%s'", projectRef.Identifier)
-		}
-		if lastGoodVersion != nil {
-			// for new repositories, we don't want to error out when we don't have
-			// any versions stored in the database so we default to the skeletal
-			// information we already have from the project file on disk
-			err = LoadProjectInto([]byte(lastGoodVersion.Config), projectRef.Identifier, project)
-			if err != nil {
-				return nil, errors.Wrapf(err, "Error loading project from "+
-					"last good version for project, %s", lastGoodVersion.Identifier)
-			}
-		} else {
-			// Check to see if there is a local configuration in the project ref
-			if projectRef.LocalConfig != "" {
-				err = LoadProjectInto([]byte(projectRef.LocalConfig), projectRef.Identifier, project)
-				if err != nil {
-					return nil, errors.Wrapf(err, "Error loading local config for project ref, %s", projectRef.Identifier)
-				}
-			}
-		}
+	lastGoodVersion, err := VersionFindOne(VersionByLastKnownGoodConfig(identifier))
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error finding recent valid version for '%s'", identifier)
 	}
-
-	if revision != "" {
-		// we immediately return an error if the repotracker version isn't found
-		// for the given project at the given revision
-		v, err := VersionFindOne(VersionByProjectIdAndRevision(projectRef.Identifier, revision))
+	if lastGoodVersion != nil {
+		err = LoadProjectInto([]byte(lastGoodVersion.Config), identifier, project)
 		if err != nil {
-			return nil, errors.Wrapf(err, "error fetching version for project %v revision %v", projectRef.Identifier, revision)
-		}
-		if v == nil {
-			// fall back to the skeletal project
-			return project, nil
-		}
-
-		project = &Project{}
-		if err = LoadProjectInto([]byte(v.Config), projectRef.Identifier, project); err != nil {
-			return nil, errors.Wrap(err, "Error loading project from version")
+			return nil, errors.Wrapf(err, "Error loading project from "+
+				"last good version for project, %s", lastGoodVersion.Identifier)
 		}
 	}
-
 	return project, nil
```

Comment:
Are we expecting LastGoodVersion to never be nil (or are we just okay with returning an unpopulated project)? Should we return an error if it is?

Reply:
My goal was to preserve the existing behavior, which was to return the minimal project in the case where the lastGoodRevision is nil. I think this would only apply when the project is brand new and there and therefore no revisions. It's not an error in that case for there not be a lastGoodRevision.